ЦЕЛЬ: Построить прогнозную модель и посмотреть, какие признаким влияют в большей мере на стоимость квартир.

In [1]:
import pandas as pd
import numpy as np

pd.set_option('display.float_format', lambda x: '%.3f' % x)

Импортим наш датасет 


In [2]:
df = pd.read_csv('good_all_data.csv')

In [3]:
from sklearn.datasets import make_regression
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import ElasticNet

In [4]:
df

Unnamed: 0             locality  rooms        price  total_area  \
0               0      Санкт-Петербург      3 13000000.000     108.000   
1               1       поселок Шушары      1  3350000.000      40.400   
2               2      Санкт-Петербург      2  5196000.000      56.000   
3               3      Санкт-Петербург      3 64900000.000     159.000   
4               4      Санкт-Петербург      2 10000000.000     100.000   
...           ...                  ...    ...          ...         ...   
23645       23645      Санкт-Петербург      3  9700000.000     133.810   
23646       23646                Тосно      3  3100000.000      59.000   
23647       23647     село Рождествено      2  2500000.000      56.700   
23648       23648      Санкт-Петербург      2 11475000.000      76.750   
23649       23649  поселок Новый Учхоз      1  1350000.000      32.300   

       living_area  kitchen_area  floor  floors_total  balcony  ...  \
0           51.000        25.000      8        16.000        0  ...   
1           18.600        11.000      1        11.000        2  ...   
2           34.300         8.300      4         5.000        0  ...   
3              NaN           NaN      9        14.000        0  ...   
4           32.000        41.000     13        14.000        0  ...   
...            ...           ...    ...           ...      ...  ...   
23645       73.300        13.830      3         5.000        0  ...   
23646       38.000         8.500      4         5.000        0  ...   
23647       29.700           NaN      1         3.000        0  ...   
23648          NaN        23.300     12        17.000        2  ...   
23649       12.300         9.000      1         5.000        0  ...   

       new_rooms  new_ceiling_height  new_ceiling_height_try  new_total_area  \
0          3.000               2.700                   2.700         108.000   
1          1.000                 NaN                     NaN          40.400   
2          2.000                 NaN                     NaN          56.000   
3          3.000                 NaN                     NaN             NaN   
4          2.000               3.030                   3.030         100.000   
...          ...                 ...                     ...             ...   
23645      3.000                 NaN                     NaN             NaN   
23646      3.000                 NaN                     NaN          59.000   
23647      2.000                 NaN                     NaN          56.700   
23648      2.000               3.000                   3.000          76.750   
23649      1.000               2.500                   2.500          32.300   

      new_days_exposition  too_fast   fast  too_slow  locality_type  \
0                     NaN     False  False     False          город   
1                  81.000     False  False     False        поселок   
2                     NaN     False  False      True          город   
3                 424.000     False  False      True          город   
4                 121.000     False  False     False          город   
...                   ...       ...    ...       ...            ...   
23645                 NaN     False  False     False          город   
23646              45.000     False  False     False          город   
23647                 NaN     False  False     False           село   
23648                 NaN     False  False      True          город   
23649                 NaN     False  False     False        поселок   

                  spb_and_lr  
0            Санкт-Петербург  
1      Ленинградская область  
2            Санкт-Петербург  
3            Санкт-Петербург  
4            Санкт-Петербург  
...                      ...  
23645        Санкт-Петербург  
23646  Ленинградская область  
23647  Ленинградская область  
23648        Санкт-Петербург  
23649  Ленинградская область  

[23650 rows x 43 columns]

Смотрим, какие колонки у нас есть

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23650 entries, 0 to 23649
Data columns (total 43 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Unnamed: 0               23650 non-null  int64  
 1   locality                 23650 non-null  object 
 2   rooms                    23650 non-null  int64  
 3   price                    23650 non-null  float64
 4   total_area               23650 non-null  float64
 5   living_area              21752 non-null  float64
 6   kitchen_area             21381 non-null  float64
 7   floor                    23650 non-null  int64  
 8   floors_total             23565 non-null  float64
 9   balcony                  23650 non-null  int64  
 10  ceiling_height           14490 non-null  float64
 11  studio                   23650 non-null  bool   
 12  apartment                23650 non-null  bool   
 13  open_plan                23650 non-null  bool   
 14  exposition_data       

Кодируем текстовые признаки с помощью Labelencoder и разбиваем выборку на тестовую и тренировочную


In [6]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder

In [7]:
# creat1ing initial dataframe
temp_df = pd.DataFrame(df[['floor_type']])

# creating instance of labelencoder
labelencoder = LabelEncoder()

# Assigning numerical values and storing in another column
df['floor_type'] = labelencoder.fit_transform(temp_df['floor_type'])

In [8]:
# creat1ing initial dataframe
temp_df = pd.DataFrame(df[['spb_and_lr']])

# creating instance of labelencoder
labelencoder = LabelEncoder()

# Assigning numerical values and storing in another column
df['spb_and_lr'] = labelencoder.fit_transform(temp_df['spb_and_lr'])

In [9]:
# creat1ing initial dataframe
temp_df = pd.DataFrame(df[['locality']])

# creating instance of labelencoder
labelencoder = LabelEncoder()

# Assigning numerical values and storing in another column
df['locality'] = labelencoder.fit_transform(temp_df['locality'])


In [10]:
# creat1ing initial dataframe
temp_df = pd.DataFrame(df[['locality_type']])

# creating instance of labelencoder
labelencoder = LabelEncoder()

# Assigning numerical values and storing in another column
df['locality_type'] = labelencoder.fit_transform(temp_df['locality_type'])

Выбираем те столбцы, с которыми будем работать. И далее избавляемся от nan

In [11]:
X_temp = df[['new_price', 'locality', 'new_rooms', 'total_area', 'floor', 'floors_total', 'floor_type', 'locality_type', 'spb_and_lr', 'parks_in_3km_number', 'center_distance_km', 'airport_distance_km', 'ponds_in_3km_number']]
X_temp = X_temp.dropna()


#center_distance_km
#airport_distance_km
#parks_in_3km_number
#nearest_park_distance_m
#ponds_in_3km_number
#nearest_pond_distance_m

Выбираем из них признаки и столбец, который будем предсказывать

In [12]:
X = X_temp[['locality', 'new_rooms', 'total_area', 'floor', 'floors_total', 'floor_type', 'locality_type', 'spb_and_lr', 'parks_in_3km_number', 'center_distance_km', 'airport_distance_km', 'ponds_in_3km_number']]
y = X_temp[['new_price']]

Создаём тестовую и тренировочную выборки

In [13]:
X_train, X_test, Y_train, Y_test = train_test_split(X, y, test_size = 0.2, random_state=42)


# Посмотрим каков у нас оказался размер объектов в трейне и тесте
print(X_train.shape)
print(X_test.shape)
print(Y_train.shape)
print(Y_test.shape)


(12828, 12)
(3208, 12)
(12828, 1)
(3208, 1)


Создаём и обучаем модель

In [14]:
lin_model = LinearRegression()
lin_model.fit(X_train, Y_train)

LinearRegression()

In [15]:
# посмотрим, какие из наших признаков оказались наиболее значимыми для определения цены на квартиру
importance = lin_model.coef_

print(importance)

[[    8244.39265448  -450362.52809194    95493.85869002    19056.92472264
     19695.91021179  -123667.96569638 -1542051.95959614  -934932.0973891
     64966.99268662   -80287.68267715     6994.28045874    99410.78500643]]


Импортируем штуки для подсчёта ошибки

In [16]:
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score

In [17]:
# Определим функцию оценки ошибки получившейся модели:

def model_assess(lin_model, X_train, X_test, Y_train, Y_test):

    # model evaluation for training set
    y_train_predict = lin_model.predict(X_train)
    rmse = (np.sqrt(mean_squared_error(Y_train, y_train_predict)))
    r2 = r2_score(Y_train, y_train_predict)

    
    print("\n")
    print("The model performance for training set")
    print("--------------------------------------")
    print('RMSE is {}'.format(rmse))
    print('R2 score is {}'.format(r2))
    print("--------------------------------------")
    print("\n")

    
    # model evaluation for testing set
    y_test_predict = lin_model.predict(X_test)
    rmse = (np.sqrt(mean_squared_error(Y_test, y_test_predict)))
    r2 = r2_score(Y_test, y_test_predict)

    print("The model performance for testing set")
    print("--------------------------------------")
    print('RMSE is {}'.format(rmse))
    print('R2 score is {}'.format(r2))

In [18]:
model_assess(lin_model, X_train, X_test, Y_train, Y_test)



The model performance for training set
--------------------------------------
RMSE is 1115580.9176451312
R2 score is 0.712433833867945
--------------------------------------


The model performance for testing set
--------------------------------------
RMSE is 1121905.2012119235
R2 score is 0.7109022739909296


Смотрим самые значимые фичи 

In [19]:
importance = lin_model.coef_[0]
importance = sorted(importance)
for i, v in enumerate(importance):
    print('Feature: %s, Score: %.5f' % (X.columns[i],v))


Feature: locality, Score: -1542051.95960
Feature: new_rooms, Score: -934932.09739
Feature: total_area, Score: -450362.52809
Feature: floor, Score: -123667.96570
Feature: floors_total, Score: -80287.68268
Feature: floor_type, Score: 6994.28046
Feature: locality_type, Score: 8244.39265
Feature: spb_and_lr, Score: 19056.92472
Feature: parks_in_3km_number, Score: 19695.91021
Feature: center_distance_km, Score: 64966.99269
Feature: airport_distance_km, Score: 95493.85869
Feature: ponds_in_3km_number, Score: 99410.78501


In [20]:
from sklearn.preprocessing import MinMaxScaler

In [21]:
def scaling(df):
    scaler = MinMaxScaler() 
    data_scaled = scaler.fit_transform(df)
    df_scaled = pd.DataFrame(data_scaled)
    return df_scaled

In [22]:
X_scaled = scaling(X)

X_train, X_test, Y_train, Y_test = train_test_split(X_scaled, y, test_size = 0.2, random_state=42)

lin_model = LinearRegression()
lin_model.fit(X_train, Y_train)

importance = lin_model.coef_
print(importance)

[[ 2094075.73423862 -2702175.16855162 16997906.84682298   609821.59112449
   1162058.70249528  -247335.93139276 -3084103.91919227  -934932.0973891
    194900.97805986 -5279718.01284927   593597.58825285   298232.35501929]]


Как видим, значения фич изменились, но сама модель особо не изменилась

In [23]:
importance = lin_model.coef_[0]
importance = sorted(importance)
for i, v in enumerate(importance):
    print('Feature: %s, Score: %.5f' % (X.columns[i],v))


Feature: locality, Score: -5279718.01285
Feature: new_rooms, Score: -3084103.91919
Feature: total_area, Score: -2702175.16855
Feature: floor, Score: -934932.09739
Feature: floors_total, Score: -247335.93139
Feature: floor_type, Score: 194900.97806
Feature: locality_type, Score: 298232.35502
Feature: spb_and_lr, Score: 593597.58825
Feature: parks_in_3km_number, Score: 609821.59112
Feature: center_distance_km, Score: 1162058.70250
Feature: airport_distance_km, Score: 2094075.73424
Feature: ponds_in_3km_number, Score: 16997906.84682


In [24]:
model_assess(lin_model, X_train, X_test, Y_train, Y_test)



The model performance for training set
--------------------------------------
RMSE is 1115580.9176451312
R2 score is 0.712433833867945
--------------------------------------


The model performance for testing set
--------------------------------------
RMSE is 1121905.201211924
R2 score is 0.7109022739909294


Пробуем новый алгоритм

In [25]:
X_train, X_test, Y_train, Y_test = train_test_split(X, y, test_size = 0.2, random_state=42)

lin_model_1 = ElasticNet()
lin_model_1.fit(X_train, Y_train)

importance = lin_model_1.coef_
print(importance)
model_assess(lin_model_1, X_train, X_test, Y_train, Y_test)

[  -3677.96858778 -170892.09328573   85747.24941514   17181.41407183
   23452.96029302  -55003.19877485   -6935.7443967   -43744.70474058
   59354.18402959  -58756.84025414    4663.66497595   84817.71726489]


The model performance for training set
--------------------------------------
RMSE is 1138814.1217186244
R2 score is 0.7003313430772117
--------------------------------------


The model performance for testing set
--------------------------------------
RMSE is 1140932.686383063
R2 score is 0.7010129367649098


In [26]:
importance = lin_model_1.coef_
importance = sorted(importance)
for i, v in enumerate(importance):
    print('Feature: %s, Score: %.5f' % (X.columns[i],v))


Feature: locality, Score: -170892.09329
Feature: new_rooms, Score: -58756.84025
Feature: total_area, Score: -55003.19877
Feature: floor, Score: -43744.70474
Feature: floors_total, Score: -6935.74440
Feature: floor_type, Score: -3677.96859
Feature: locality_type, Score: 4663.66498
Feature: spb_and_lr, Score: 17181.41407
Feature: parks_in_3km_number, Score: 23452.96029
Feature: center_distance_km, Score: 59354.18403
Feature: airport_distance_km, Score: 84817.71726
Feature: ponds_in_3km_number, Score: 85747.24942


ВЫВОД: В итоге удалось получить результат RMSE is 1140932,R2 score is 0.701 для тестовой выборки. Наиболее значимые признаки оказались lacality, new rooms, center distance,airport distance. На основе этих данных можно сделать вывод, что стоимость квартиры в большей степени зависит от расположения, количества квартир, расстояния до аэропорта.
